### PR TITLE
[v13] Use pnpm for package management

### PR DESCRIPTION
**Pull request by ryanwilsonperkin**
_Edited by @jesstelford_: Because _speed_.

### CODE REVIEW

**Comment by alex-page** [comment]
If we are going to use pnpm workspaces we should remove all the other stuff like `package.json` workspaces and turborepo.

**Comment by ryanwilsonperkin** [comment]
A whole bunch of these commits could be shipped independently if we'd like to reduce the size of this PR. They're fixing issues that only become apparent after pnpm's change to no longer hoist transitive & sub-package dependencies, and would work without pnpm

**Review by sam-b-rose — APPROVED** [review]
Tested and all tasks and workflow appear to be functioning properly 👍 

Can we also remove the `workspaces` key with `packages and `nohoist`?

Screenshot 2024-03-28 at 1 26 27 PM

File: polaris-react/config/postcss-plugins.js
Comment (jesstelford) [inline polaris-react/config/postcss-plugins.js:13]: Is `pnpm` symlinking `@shopify/polaris-tokens` to `../../node_modules/@shopify/polaris-tokens` here when doing local dev / build?

File: package.json
Comment (jesstelford) [inline package.json:76]: I see this is now moved into the `polaris.shopify.com` workspace (nice!) - is that because pnpm installs a full suite of deps into `node_modules` for _every_ workspace as symlinks (unlike yarn which hoists them to the parent `node_modules` where possible)?

File: .github/workflows/changelog.yml
Comment (alex-page) [inline .github/workflows/changelog.yml]: Can we put the pnpm version into the package.json `packageManager`. I don't want to manage versions hard coded across GitHub actions. The version should be centralised.

File: polaris-react/package.json
Comment (jesstelford) [inline polaris-react/package.json:89]: Are these missing deps that _happened_ to be available because of yarn's hoisting, but now need to be explicitly installed per workspace so pnpm can symlink them correctly? If so; nice!

File: polaris-react/src/components/ThemeProvider/ThemeProvider.tsx
Comment (jesstelford) [inline polaris-react/src/components/ThemeProvider/ThemeProvider.tsx:19]: Bunch of these appear to be prettier-related changes.

Do you mind if we revert these and bring them in a separate PR (they're [functionally equivalent](https://www.typescriptlang.org/play?ts=4.7.2#code/KYDwDg9gTgLgBAYwgOwM7xgC2AW2AOQEM9UAZCBQgGzgF44BtAcioEsBzTGJgGjiYAmhKAGsAtKDDAorPMhjUmAXTiFUiFOgDcAWABQ+mAE8pcACrY8RPOUo16xqRABmcLLgLFgZCtQbIAVxwAI2klXT0Aeki4WLgAPQB+QxNgc0tPG197OAAKR2AXNwzrb1tqAEp-INCocP1ouITEoA))?

File: polaris.shopify.com/scripts/get-props/yarn.lock
Comment (jesstelford) [inline polaris.shopify.com/scripts/get-props/yarn.lock:1]: `polaris.shopify.com/scripts/get-props` has [its own `package.json`](https://github.com/Shopify/polaris/blob/main/polaris.shopify.com/scripts/get-props/package.json) (for reasons I'm not entirely certain, but I _think_ it's because it needed a version of a dep that's incompatible with the workspace's version), how will `pnpm` behave in this situation? Do we need to pull the script up into its own workspace so it can perform the same function?

**Review by jesstelford — APPROVED** [review]
Approved! With one ask to revert some of the code formatting changes (just to keep the diff size down).

File: .github/workflows/changelog.yml
Comment (alex-page) [inline .github/workflows/changelog.yml]: Any reason we are using v2 and not v3?

**Comment by jesstelford** [comment]
I hacked together a script (below) to try compare the build output with pnpm expecting zero changes (it's the package _manager_ that's changed, not the _packages themselves_), but... There's quite a bit of difference 😓

Some of it seems to be slight differences in however the build tool has generated the final JS with extra parens here and there, but then there's other parts which appear to change the exports of files, etc 😱 

Give it a run yourself and see if you're noticing similar differences:

Put this in the root of the Polaris repo as `script.sh`, allow it to be executed with `chmod a+x ./script.sh`, then run it `./script.sh`

```sh
#!/usr/bin/env bash
declare -a build_folders=(
  "polaris-for-vscode/dist/"
  "polaris-icons/dist/"
  "polaris-migrator/dist/"
  "polaris-react/build/"
  "polaris-tokens/dist/"
  "polaris.shopify.com/.next/"
  "polaris.shopify.com/public/playroom/"
  "polaris.shopify.com/public/og-images/"
  "polaris.shopify.com/public/og-images/"
)

git fetch
source ~/.nvm/nvm.sh
MAIN_COMMIT=`git merge-base origin/pnpm origin/main`

# # Get a clean repo at the latest shared ancestor that's using yarn
git worktree add ../polaris-main $MAIN_COMMIT
cd ../polaris-main
nvm use
yarn
yarn build
for i in "${build_folders[@]}"
do
  mkdir -p "../polaris-build/main/""$i"
  cp -R "$i" "../polaris-build/main/""$i"
done

cd -

# Cleanup
git worktree remove -f ../polaris-main

git worktree add ../polaris-pnpm origin/pnpm
cd ../polaris-pnpm
nvm use
pnpm install
pnpm build
for i in "${build_folders[@]}"
do
  mkdir -p "../polaris-build/pnpm/""$i"
  cp -R "$i" "../polaris-build/pnpm/""$i"
done

cd -
# Cleanup
git worktree remove -f ../polaris-pnpm

echo "Now run: git diff --no-index ../polaris-build/main ../polaris-build/pnpm"
```

File: .github/workflows/changelog.yml
Comment (ryanwilsonperkin) [reply .github/workflows/changelog.yml]: I just copied this from their docs which probably referenced the old version, can fix

File: polaris-react/src/components/ThemeProvider/ThemeProvider.tsx
Comment (ryanwilsonperkin) [reply polaris-react/src/components/ThemeProvider/ThemeProvider.tsx:19]: Works for me!

File: polaris-react/package.json
Comment (ryanwilsonperkin) [reply polaris-react/package.json:89]: That's right! They were being used directly but weren't being installed explicitly, and began failing because pnpm won't hoist them and calls them "ghost dependencies"

File: .github/workflows/changelog.yml
Comment (ryanwilsonperkin) [reply .github/workflows/changelog.yml]: Great idea

File: package.json
Comment (ryanwilsonperkin) [reply package.json:76]: Yes, it'll create the top-level `node_modules` with things needed at the root of the workspace, and then a separate one for each sub-package, using just the dependencies that those declare. This was a case of the `polaris.shopify.com` package needing `patch-package` and the `patches/` directory for its own case, but it had been erroneously put at the top-level instead

File: polaris-react/config/postcss-plugins.js
Comment (ryanwilsonperkin) [reply polaris-react/config/postcss-plugins.js:13]: In fact, `require.resolve` will resolve to the actual path to the `dist/` directory here. I think that might be `turbo` setting that up, though it could be `pnpm`. Previously it would've been building the polaris-tokens and linking it into node_modules, but now its smart enough to point directly to the build files

File: polaris.shopify.com/scripts/get-props/yarn.lock
Comment (ryanwilsonperkin) [reply polaris.shopify.com/scripts/get-props/yarn.lock:1]: It seemed that the `get-props` script was still able to run after switching package management over to PNPM so I opted to remove it. Perhaps we should remove the `package.json` as well?

**Comment by ryanwilsonperkin** [comment]
> I hacked together a script (below) to try compare the build output with pnpm expecting zero changes (it's the package _manager_ that's changed, not the _packages themselves_), but... There's quite a bit of difference 😓
> 
> Some of it seems to be slight differences in however the build tool has generated the final JS with extra parens here and there, but then there's other parts which appear to change the exports of files, etc 😱

Thanks for putting this together @jesstelford ! It looks to me like a lot of changes to the nextJS compiled output (shouldn't be a big deal for our purposes) and also some changes to the way the dist files get build & minified. This suggests to me that perhaps the version of Rollup or one of its sub-dependencies has changed. Maybe a version that one of the packages depends on was previously improperly hoisted?

It'd be great to try testing the bundles in some of our consumers like `Shopify/Web` and see if they result in any differences (they shouldn't)

File: polaris-icons/package.json
Comment (aaronccasanova) [inline polaris-icons/package.json]: Any insights on why this was introduced?

Comment (ryanwilsonperkin) [reply polaris-icons/package.json]: `@shopify/polaris-icons` produces a build that depends on React as a peer dependency but was not declaring it. Take for example this file that it exports: https://www.unpkg.com/browse/@shopify/polaris-icons@8.9.0/dist/icons/AdjustIcon.svg.mjs

```js
import React from 'react';

var SvgAdjustIcon = function SvgAdjustIcon(props) {
  return /*#__PURE__*/React.createElement("svg", Object.assign({
    viewBox: "0 0 20 20"
  }, props), /*#__PURE__*/React.createElement("path", {
    fillRule: "evenodd",
    d: "M9.095 6.25a3.001 3.001 0 0 1 5.81 0h1.345a.75.75 0 0 1 0 1.5h-1.345a3.001 3.001 0 0 1-5.81 0h-5.345a.75.75 0 0 1 0-1.5h5.345Zm1.405.75a1.5 1.5 0 1 1 3 0 1.5 1.5 0 0 1-3 0Z"
  }), /*#__PURE__*/React.createElement("path", {
    fillRule: "evenodd",
    d: "M8 16a3.001 3.001 0 0 0 2.905-2.25h5.345a.75.75 0 0 0 0-1.5h-5.345a3.001 3.001 0 0 0-5.81 0h-1.345a.75.75 0 0 0 0 1.5h1.345a3.001 3.001 0 0 0 2.905 2.25Zm1.5-3a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0Z"
  }));
};
SvgAdjustIcon.displayName = "AdjustIcon";

export { SvgAdjustIcon as S };
```

It imports `react` but did not specify it as a peer dependency, which meant that under pnpm it would not have access to it for producing its build and resulted in errors when linking it.

This can be shipped separately from this PR if you'd like me to split it out, its an issue with the current state of the package, pnpm just made it more apparent

Comment (ryanwilsonperkin) [reply polaris-icons/package.json]: Extracting to a separate PR here: https://github.com/Shopify/polaris/pull/11825

File: polaris-react/src/components/ThemeProvider/ThemeProvider.tsx
Comment (ryanwilsonperkin) [reply polaris-react/src/components/ThemeProvider/ThemeProvider.tsx:19]: Extracted here: https://github.com/Shopify/polaris/pull/11824

Comment (ryanwilsonperkin) [reply polaris-react/src/components/ThemeProvider/ThemeProvider.tsx:19]: Interesting, CI fails there: https://github.com/Shopify/polaris/actions/runs/8528174905/job/23361209192?pr=11824
So it would seem that we're ending up with a different version of a key package (TypeScript? Eslint?) when linting one of these packages if using yarn instead of pnpm

File: polaris-icons/package.json
Comment (aaronccasanova) [reply polaris-icons/package.json]: I see! I guess the challenge is we have some consumers of `@shopify/polaris-icons` that only use the SVGs (not the React components). Perhaps, this is a good use case for [`optionalDependencies`](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#optionaldependencies)?

Comment (ryanwilsonperkin) [reply polaris-icons/package.json]: Sam's got a comment on the PR about using peerDependenciesMeta to mark them as optional, we could potentially use that approach but couldn't use `optionalDependencies` because pnpm won't install those as it would peers, so they won't be accessible to `@shopify/polaris-icons` to import (would fail for any application that tried to install `@shopify/polaris-icons` and use the React icons)

File: polaris.shopify.com/scripts/get-props/yarn.lock
Comment (jesstelford) [reply polaris.shopify.com/scripts/get-props/yarn.lock:1]: Yeah, let's do it! Cleaning up is always better than perpetuating workarounds 🎉

File: polaris-react/src/components/ThemeProvider/ThemeProvider.tsx
Comment (jesstelford) [reply polaris-react/src/components/ThemeProvider/ThemeProvider.tsx:19]: Ah damn, in that case, leave the changes in this PR. Sorry for the extra work there!

**Comment by ryanwilsonperkin** [comment]
@sam-b-rose after removing the `.workspaces` reference from package.json the linter is failing because `.eslintrc.js` was relying on it to find package directories:

```js
const packages = require('./package.json').workspaces.packages;
```

If we want a 1-1 replacement, I can add a dependency on [`js-yaml`](https://www.npmjs.com/package/js-yaml) to the workspace root so that we can read the `pnpm-workspace.yaml` file to retrieve this information. Thoughts?

File: polaris.shopify.com/scripts/get-props/yarn.lock
Comment (ryanwilsonperkin) [reply polaris.shopify.com/scripts/get-props/yarn.lock:1]: Looked at this a bit more today, it looks like the reason for it having a package.json is because it uses dependencies that the `polaris.shopify.com` package doesn't declare (jest, babel) and adds a test script. Since those would otherwise be hoisted to the root, it does need to preserve the package.json. Two options here, let me know your preference:

1. Lift the `test` script and dependencies up to `polaris.shopify.com/package.json`
2. Add a new workspace project for `polaris.shopify.com/scripts/get-props`

I prefer the first option, but am happy to accept either

**Comment by aaronccasanova** [comment]
@ryanwilsonperkin +1 to continuing with `pnpm-workspace.yaml` as the source of truth and introducing `js-yaml` for feature parity in `.eslintrc.js`

**Comment by jesstelford** [comment]
@ryanwilsonperkin We're prepping for a major release of Polaris in the next week or so ([to update the minimum node version](https://github.com/Shopify/polaris-internal/issues/1524)) - how do you feel about this change being merged as part of the major release?

I'm happy with the change, but given the changes seen in that script I ran, I wanted to just play it safe and push it out as a major in case something has changed.

☝️  Moderate opinion, weakly held

**Review by aaronccasanova — APPROVED** [review]

**Comment by alex-page** [comment]
@jesstelford I like that idea as it's a pretty significant change for developers contributing to Polaris.

**Comment by jesstelford** [comment]
Ok, in that case I'll mark this PR as Draft until we're ready for the v13 release.

@ryanwilsonperkin are you ok if myself & @aaronccasanova take over this PR from here? We'll shepherd it into the wild ASAP 🎉

**Comment by ryanwilsonperkin** [comment]
Sounds good to me! Thanks @jesstelford

File: polaris.shopify.com/scripts/get-props/yarn.lock
Comment (jesstelford) [reply polaris.shopify.com/scripts/get-props/yarn.lock:1]: Yep, #1 sounds like the correct approach here - sounds like it was a workaround due to yarn's hoisting. pnpm doesn't have that problem, so let's just push it up to the `polaris.shopify.com/package.json` :+1:

**Comment by jesstelford** [comment]
Heads up; I've rebased this on top of the `node-20` branch (PR: https://github.com/Shopify/polaris/pull/11844) so we can ensure they're playing together nicely, and we can merge this PR immediately after `node-20` is merged 👍

File: polaris.shopify.com/scripts/get-props/yarn.lock
Comment (aaronccasanova) [reply polaris.shopify.com/scripts/get-props/yarn.lock:1]: Agree on moving forward with option 1 👍

File: package.json
Comment (aaronccasanova) [inline package.json]:
```suggestion
    "new-migration": "pnpm --filter @shopify/polaris-migrator generate"
```

File: polaris-react/src/components/AppProvider/global.css
Comment (jesstelford) [inline polaris-react/src/components/AppProvider/global.css:5]: `yarn` used to hoist everything to the root `node_modules`, but `pnpm` doesn't hoist like that, so we only need to go up to the `polaris-react/node_modules` to find the dependency.

File: polaris.shopify.com/scripts/get-props/yarn.lock
Comment (jesstelford) [reply polaris.shopify.com/scripts/get-props/yarn.lock:1]: Option 1 didn't work 😓 

Have reverted to a nested workspace and pulled this out into an issue to tackle later: https://github.com/Shopify/polaris/issues/11863

**Comment by ryanwilsonperkin** [comment]
🚀🚀🚀

**Pull request merged**
